model/span/event.go

```diff
@@ -447,5 +447,37 @@ func (e *Event) fields(tctx *transform.Context) common.MapStr {
 
 	st := e.Stacktrace.Transform(tctx)
 	utility.Set(fields, "stacktrace", st)
+
+	if fingerprint := e.servicemapFingerprint(tctx); fingerprint != "" {
+		utility.Set(fields, "servicemap", common.MapStr{"fingerprint": fingerprint})
+	}
 	return fields
 }
+
+func (e *Event) servicemapFingerprint(tctx *transform.Context) string {
+	if e.Destination == nil || e.Destination.Address == nil || *e.Destination.Address == "" {
+		return ""
+	}
+	var buf bytes.Buffer
+	buf.WriteString(*e.Destination.Address)
+
+	add := func(s string) {
+		buf.WriteString("|")
+		buf.WriteString(s)
+	}
+	add(e.Type)
+	if e.Subtype != nil {
+		add(*e.Subtype)
+	}
+	if e.Service != nil && e.Service.Name != nil {
+		add(*e.Service.Name)
+	} else if tctx.Metadata.Service != nil && tctx.Metadata.Service.Name != nil {
+		add(*tctx.Metadata.Service.Name)
+	}
+	if e.Service != nil && e.Service.Environment != nil {
+		add(*e.Service.Environment)
+	} else if tctx.Metadata.Service != nil && tctx.Metadata.Service.Environment != nil {
+		add(*tctx.Metadata.Service.Environment)
+	}
+	return base64.StdEncoding.EncodeToString(buf.Bytes())
```

Comment:
What's the reason behind base64-encoding? I can see that it would be necessary if there were any binary values involved, but there aren't any.

I also don't understand why we're not hashing - that would produce smaller strings. By using a fast, non-cryptographic hash like xxhash64, overhead should be pretty minimal, will avoid allocations and copying, and results in a compact value.

Reply:
This is a valid point, I changed the logic to make use of the xxhash algorithm, which was leading to only one additional allocation per event, which is negligible, and comparable processing time. 

The logic is now basically exactly what is done in the [fingerprint processor](https://github.com/elastic/beats/blob/master/libbeat/processors/fingerprint/fingerprint.go) in beats. I still went with implementing it directly in the transformation function to avoid the need of overriding the [MakeDefaultObserverSupport](https://github.com/elastic/beats/blob/master/libbeat/publisher/processing/default.go#L85) for ensuring to always set the processor in code.Some logic would need to be exported, and I think the current code is simple enough to duplicate instead of making other required changes.